kong/clustering/services/sync/rpc.lua

```diff
@@ -173,31 +176,133 @@ local function is_rpc_ready()
 end
 
 
-local function do_sync()
+local function lmdb_purge()
+  local t = assert(txn.begin(512))
+  t:db_drop(false)
+  t:set(DECLARATIVE_HASH_KEY, DECLARATIVE_EMPTY_CONFIG_HASH)
+  assert(t:commit()) -- not much we can do if this fails
+
+  -- we are at a unready state
+  -- consider the config empty
+  kong.core_cache:purge()
+  kong.cache:purge()
+end
+
+
+local function reset_dp()
+  ngx_log(ngx_INFO, "resetting the data plane")
+
+  lmdb_purge()
+  -- the `init` is a magic version that results in an empty router unconditionally
+  assert(build_router("init"))
+end
+
+
+local do_sync
+
+
+local function error_handler(err, is_full_sync)
+  ngx_log(ngx_ERR, err)
+
+  if is_full_sync then
+    -- in case of full sync, reset the DP to initial state and retry immediately
+    reset_dp()
+    return do_sync() -- retry without releasing the mutex as the DP has nothing to do except syncing
+  end
+
+  return nil, err
+end
+
+
+local function lmdb_upsert(db, t, delta, opts)
+  local delta_type = delta.type
+  local delta_entity = delta.entity
+
+  -- upsert the entity
+  -- does the entity already exists?
+  local old_entity, err = db[delta_type]:select(delta_entity)
+
+  if err then
+    return nil, err
+  end
+
+  if old_entity then
+    local res, err = delete_entity_for_txn(t, delta_type, old_entity, opts)
+    if not res then
+      return nil, err
+    end
+  end
+
+  local res, err = insert_entity_for_txn(t, delta_type, delta_entity, opts)
+  if not res then
+    return nil, err
+  end
+
+  return { delta_type, old_entity and "update" or "create", delta_entity, old_entity, }
+end
+
+
+local function lmdb_delete(db, t, delta, opts, is_full_sync)
+  local delta_type = delta.type
+
+  -- NOTE: the first page of full sync MUST NOT contain delete events
+  -- assert(not t.purged, "unexpected delete event in frist page of full sync")
+
+  local old_entity, err = db[delta_type]:select(delta.pk, opts)
+  if err then
+    return nil, err
+  end
+
+  -- entity does not exist, no need to delete
+  if not old_entity then
+    return true
+  end
+
+  -- delete the entity
+  local res, err = delete_entity_for_txn(t, delta_type, old_entity, opts)
+  
+  if is_full_sync then
+    local entity_not_found = err and err:match("MDB_NOTFOUND")
+    -- full sync needs extra tolerance for errors
+    if not res and not entity_not_found then
+      return nil, err
+    end
+  else
+    if not res then
+      return nil, err
+    end
+  end
```

Comment:
Not a block, i know you're struggling with this PR.

```suggestion
  if is_full_sync then
    local entity_not_found = err and err:match("MDB_NOTFOUND")
    -- full sync needs extra tolerance for errors
    if not res and not entity_not_found then
      return nil, err
    end
  end

  if not res then
    return nil, err
  end
```

Reply:
They are different. The later is incorrect.